internal/mock/storage.go

```diff
@@ -0,0 +1,100 @@
+package mock
+
+import (
+	"github.com/gnolang/gno/tm2/pkg/bft/types"
+
+	"github.com/gnolang/tx-indexer/storage"
+)
+
+var _ storage.Storage = &Storage{}
+
+type Storage struct {
+	GetLatestSavedHeightFn func() (int64, error)
+	GetWriteBatchFn        func() storage.Batch
+	GetBlockFn             func(int64) (*types.Block, error)
+	GetTxFn                func([]byte) (*types.TxResult, error)
+}
+
+func (m *Storage) GetLatestHeight() (int64, error) {
+	if m.GetLatestSavedHeightFn != nil {
+		return m.GetLatestSavedHeightFn()
+	}
+
+	return 0, nil
+}
+
+// GetBlock fetches the block by its number
+func (m *Storage) GetBlock(blockNum int64) (*types.Block, error) {
+	if m.GetBlockFn != nil {
+		return m.GetBlockFn(blockNum)
+	}
+
+	panic("not implemented")
+}
+
+// GetTx fetches the tx using its hash
+func (m *Storage) GetTx(tx []byte) (*types.TxResult, error) {
+	if m.GetTxFn != nil {
+		return m.GetTxFn(tx)
+	}
+
+	panic("not implemented")
+}
+
+// WriteBatch provides a batch intended to do a write action that
+// can be cancelled or committed all at the same time
+func (m *Storage) WriteBatch() storage.Batch {
+	if m.GetWriteBatchFn != nil {
+		return m.GetWriteBatchFn()
+	}
+
+	panic("not implemented")
+}
+
```

Comment:
The preference we have for unimplemented mock methods is to return nil / empty object, instead of panicking

Please make sure these `Storage` mock overrides don't panic, but return nil / empty 🙏

Reply:
I tried to follow actual preferences, but I did not in that case because:

- WriteBatch() will never return a nil Batch, so I wanted to panic as a warning to the dev using the mocks.
- The same case on GetTx and GetBlock, they never will be nil (they will be ErrNotFound or a result)

Ping me if you think I should remove the panics, maybe I'm not understanding some use case. Thanks!

Reply:
<img src="https://media1.giphy.com/media/4ZepGNwhW4AaA/giphy.gif"/>